specs/lifespan.rst

```diff
@@ -0,0 +1,111 @@
+=================
+Lifespan Protocol
+=================
+
+**Version**: 1.0 (2018-09-06)
+
+The Lifespan ASGI sub-specification outlines how to communicate
+lifespan events such as startup and cleanup within ASGI. The lifespan
+being referred to is that of main event loop. In a multi-process
+environment there will be lifespan events in each process.
+
+The lifespan messages allow for a application to initialise and
+cleanup in the context of a running event loop. An example of this
+would be creating a connection pool and subsequently closing the
+connection pool to release the connections.
+
+A possible implementation of this protocol is given below::
+
+    class App:
+
+        def __init__(self, scope):
+            self.scope = scope
+
+        async def __call__(self, receive, send):
+            if self.scope['type'] == 'lifespan':
+                while True:
+                    message = await receive()
+                    if message['type'] == 'lifespan.startup':
+                        await self.startup()
+                        await send({'type': 'lifespan.startup.complete'})
+                    elif message['type'] == 'lifespan.cleanup':
+                        await self.cleanup()
+                        await send({'type': 'lifespan.cleanup.complete'})
+                        return
+            else:
+                pass # Handle other types
+
+        async def startup(self):
+            ...
+
+        async def cleanup(self):
+            ...
+
+
+Scope
+'''''
+
+The lifespan scope exists for the duration of the event loop. The
+scope itself contains basic metadata,
+
+* ``type``: ``lifespan``
+
+Servers must ignore any exceptions raised when calling the application
```

Comment:
I'd add a comment further saying that if an application errors with the lifespan protocol then the server should explicitly continue but not run lifespan events, to make it crystal clear.